ocaml/xapi/certificates.ml

```diff
@@ -143,6 +153,86 @@ let raise_corrupt kind n =
 let raise_library_corrupt () =
   raise (Server_error (certificate_library_corrupt, []))
 
+module Db_util : sig
+  type name = string
+
+  val add_cert :
+       __context:Context.t
+    -> type':[< `host of API.ref_host | `ca of name]
+    -> X509.Certificate.t
+    -> API.ref_Certificate
+
+  val remove_cert_by_ref : __context:Context.t -> API.ref_Certificate -> unit
+
+  val remove_ca_cert_by_name : __context:Context.t -> name -> unit
+
+  val get_host_certs :
+    __context:Context.t -> host:API.ref_host -> API.ref_Certificate list
+  (** [get_host_certs ~__context ~host] gets all the host certs in the database
+    * belonging to [host] (the term 'host' is overloaded here) *)
+end = struct
+  module Date = Xapi_stdext_date.Date
+
+  type name = string
+
+  let add_cert ~__context ~type' certificate =
+    let name, host, _type =
+      match type' with
+      | `host host ->
+          ("", host, `host)
+      | `ca name ->
+          (name, Ref.null, `ca)
+    in
+    let date_of_ptime time = Date.of_float (Ptime.to_float_s time) in
+    let dates_of_ptimes (a, b) = (date_of_ptime a, date_of_ptime b) in
+    let not_before, not_after =
+      dates_of_ptimes (X509.Certificate.validity certificate)
+    in
+    let fingerprint =
+      X509.Certificate.fingerprint Mirage_crypto.Hash.(`SHA256) certificate
+      |> pp_hash
+    in
+    let uuid = Uuid.(to_string (make_uuid ())) in
+    let ref' = Ref.make () in
+    Db.Certificate.create ~__context ~ref:ref' ~uuid ~host ~not_before
+      ~not_after ~fingerprint ~name ~_type ;
+    ref'
+
+  let get_host_certs ~__context ~host =
+    let open Db_filter_types in
+    let type' = Eq (Field "type", Literal "host") in
+    let host' = Eq (Field "host", Literal (Ref.string_of host)) in
+    let expr = And (type', host') in
+    Db.Certificate.get_refs_where ~__context ~expr
+
+  let remove_cert_by_ref ~__context self =
+    Db.Certificate.destroy ~__context ~self
+
+  let remove_ca_cert_by_name ~__context name =
+    let expr =
+      let open Db_filter_types in
+      let type' = Eq (Field "type", Literal "ca") in
+      let name' = Eq (Field "type", Literal name) in
+      And (type', name')
+    in
+    let self =
+      match Db.Certificate.get_refs_where ~__context ~expr with
+      | [x] ->
+          x
+      | xs ->
+          let ref_str =
+            xs |> List.map Ref.short_string_of |> String.concat ", "
+          in
+          D.error
+            "expected 1 certificate with name='%s', but found several: [ %s ]"
```

Comment:
This case also happens when the list is empty, not just when there's more than one.

The invalid value is a good error the return in that case.

If there is more than one returned I would say that's an internal error and the user should contact support.